src/services/dataRenderService.ts

```diff
@@ -0,0 +1,582 @@
+            }
+        }
+
+        private drawOuter(element: d3.Selection<any>) {
+            let outlineArc = d3.svg.arc()
+                .innerRadius(this.settings.outerLine.showStraightLines ? this.innerRadius : this.outerRadius)
+                .outerRadius(this.outerRadius);
+
+            let outerThickness: string = this.settings.outerLine.thickness + "px",
+                color: string = this.settings.outerLine.color;
+
+            let outerLine = element.selectAll(DataRenderService.OuterLine.selectorName).data(this.dataPoints);
+            outerLine.enter().append("path");
+            outerLine
+                .attr("fill", "none")
+                .attr({
+                    "opacity": 0.5,
+                    "stroke": color,
+                    "stroke-width": outerThickness,
+                    "d": outlineArc as SvgArc<any> // TODO: check it.
+                })
+                .style("opacity", 0.5)
+                .classed(DataRenderService.OuterLine.className, true);
+            outerLine.exit().remove();
+        }
+
+        public drawOuterLines(element: d3.Selection<any>): void {
+            let settings: AsterPlotSettings = this.settings;
+
+            this.drawOuter(element);
+
+            if (settings.outerLine.showGrid || settings.outerLine.showGridTicksValues) {
+                this.drawGrid(element, settings.outerLine);
+            } else {
+                this.cleanGrid(element);
+            }
+        }
+
+        private cleanGrid(element: d3.Selection<any>): void {
+            element.selectAll(DataRenderService.CircleLine.selectorName).remove();
+            element.selectAll(DataRenderService.CircleText.selectorName).remove();
+            element.selectAll("circle").remove();
+        }
+
+        public cleanOuterLines(element: d3.Selection<any>): void {
+            element.selectAll(DataRenderService.OuterLine.selectorName).remove();
+            this.cleanGrid(element);
+        }
+
+        private calcTickOptions(value: number): CircleTicksOptions {
+            let val = value > 0 ? Math.floor(value) : Math.ceil(value);
+            let modifier = 1;
+
+            if (val === 0) {
+                for (let i = 0; i < value.toString().length - 3; ++i) {
+                    modifier *= 10;
+                    val = value * modifier;
+                    val = val > 0 ? Math.floor(val) : Math.ceil(val);
+
+                    if (val !== 0) {
+                        break;
+                    }
+                }
+            }
+
+            let step = Math.pow(10, val.toString().length - 1);
+
+            let count = Math.ceil((val) / step),
+                endPoint = count * step / modifier,
+                diffPercent = endPoint / value;
+
+            return {
+                diffPercent,
+                maxHeight: count * step * modifier,
+                ticksCount: count % 3 === 0 ? 3 : 2 // 2 or 3 ticks only needed
```

Comment:
Too many magic numbers. Is it possible to create variables and set names fro them?